### ISSUE TITLE
Option to enable compiler sanitizers

### ISSUE THREAD

**Issue by eflumerf** [DUNE-DAQ/daq-cmake#84]
I went through a bit of pain yesterday trying to figure out how to enable the Address Sanitizer to try to track down a memcpy issue. The basic problem was that with the sanitizer enabled, Python couldn't load the pybind libraries correctly. I eventually discovered that adding the -fsanitize=address flags to the applications only (and performing a debug build) gave me the ASAN functionality for the daq applications while leaving the python libraries alone.

```
diff --git a/cmake/DAQ.cmake b/cmake/DAQ.cmake
index d4bb35e..2f69602 100644
--- a/cmake/DAQ.cmake
+++ b/cmake/DAQ.cmake
@@ -631,6 +631,9 @@ function(daq_add_application appname)
   )
   add_dependencies( ${appname} ${PRE_BUILD_STAGE_DONE_TRGT})
 
+  target_compile_options(${appname} PUBLIC -fsanitize=address)
+  target_link_options(${appname} PUBLIC -fsanitize=address)
+
   _daq_set_target_output_dirs( ${appname} ${APP_PATH} )
   _daq_define_exportname()
```

**Comment by eflumerf** [comment]
Other sanitizers of interest: -fsanitize=thread, -fsanitize=undefined

**Comment by eflumerf** [comment]
Obviously, this would be guarded by a compile option (passed in via a `dbt-build` flag?)

**Comment by jcfreeman2** [comment]
I propose that we get this in for `fddaq-v5.5.0` and this sanitization occur as part of the already existing `--debug` option.

**Comment by jcfreeman2** [comment]
On further thought, there are two arguments against just extending the use of `--debug`:
* Sanitizers can pick up different things at different optimization levels, so we wouldn't want to tie it to a particular optimization level, which is what tying it to `--debug` would do
* I'd like to support both address and thread sanitization; however, doing both at the same time doesn't work, so we can't just have `--debug` cover both use cases -- though we could have `--debug address` and `--debug thread` on top of the traditional `--debug` if we decided we didn't care about instrumenting optimized code
